pkg/sql/schemachanger/scplan/plan.go

```diff
@@ -223,132 +225,150 @@ func buildStages(init scpb.State, g *scgraph.Graph, params Params) []Stage {
 			break
 		}
 		// Sort ops based on graph dependencies.
-		sortOps(g, s.Ops.Slice())
+		SortOps(g, s.Ops.Slice())
 		stages = append(stages, s)
 		cur = s.After
 	}
 	validateStages(stages)
 	return stages
 }
 
-// Check if some route exists from curr to the
-// target node
-func doesPathExistToNode(graph *scgraph.Graph, start *scpb.Node, target *scpb.Node) bool {
-	nodesToVisit := []*scpb.Node{start}
-	visitedNodes := map[*scpb.Node]struct{}{}
-	for len(nodesToVisit) > 0 {
-		curr := nodesToVisit[0]
-		if curr == target {
-			return true
+// SortOps sorts the operations into order based on
+// graph dependencies
+func SortOps(graph *scgraph.Graph, ops []scop.Op) {
```

Comment:
I find myself surprised by this approach. I would have thought that it'd be easier to sort the edges than the ops. It seems to me like any time you need to sort, you already have the edges in scope. The edges then have the nodes and will generally avoid any need to go back and find nodes from ops.

Reply:
I believe that we can create a full topological sort of nodes at the outset of planning with the below code. We can then use that to sort op edges on their `To` node. I think that'll do it. Am I missing something? Consider the below code to generate a topological sort order:

```go
// rankNodes performs a topological sort of the nodes in the graph.
func rankNodes(g *Graph) map[*scpb.Node]int {
	backCycleExists := func(n *scpb.Node, de *DepEdge) bool {
		var foundBack bool
		_ = g.ForEachDepEdgeFrom(de.To(), func(maybeBack *DepEdge) error {
			foundBack = foundBack || maybeBack.To() == n
			return nil
		})
		return foundBack
	}
	l := list.New()
	marks := make(map[*scpb.Node]bool)
	var visit func(n *scpb.Node)
	visit = func(n *scpb.Node) {
		permanent, marked := marks[n]
		fmt.Println(marked, permanent, screl.NodeString(n))
		if marked && !permanent {
			panic("not a dag")
		}
		if marked && permanent {
			return
		}
		marks[n] = false
		g.ForEachDepEdgeFrom(n, func(de *DepEdge) error {
			// We want to eliminate cycles caused by swaps. In that
			// case, we want to pretend that there is no edge from the
			// add to the drop, and, in that way, the drop is ordered first.
			if n.Direction != scpb.Target_ADD || !backCycleExists(n, de) {
				visit(de.To())
			}
			return nil
		})
		if oe, ok := g.GetOpEdgeFrom(n); ok {
			visit(oe.To())
		}
		marks[n] = true
		l.PushFront(n)
	}

	_ = g.ForEachNode(func(n *scpb.Node) error {
		visit(n)
		return nil
	})
	rank := make(map[*scpb.Node]int, l.Len())
	for i, cur := 0, l.Front(); i < l.Len(); i++ {
		rank[cur.Value.(*scpb.Node)] = i
		cur = cur.Next()
	}
	return rank
}
```